### ISSUE TITLE
Zap message history tag not cover all json scenarios 

### ISSUE THREAD

**Issue by roisec** [zaproxy/zaproxy#8399]
### Is your feature request related to a problem? Please describe.

Hi 
I encourage some use cases that http request contains json requests, and there is no tag on the relevant message.
I think the tagging mechanism only works on HTTP responses.

### Describe the solution you'd like

Cover all the http requests with json 

### Describe alternatives you've considered

Filter out by myself .

### Would you like to help fix this issue?

- [ ] Yes

**Comment by thc202** [comment]
The tagging mechanism works on requests and responses but the default tag is only for responses (as the name indicates `response_json`).

**Comment by roisec** [comment]
I saw in history some HTTP requests and responses without the JSON tag, but they contained the header content-type: application/JSON.
Would you happen to know why?

**Comment by kingthorin** [comment]
Tagging is based on passive scan which uses a queue, so they just may not have been processed/tagged yet.

**Comment by kingthorin** [comment]
Correction the patterns are case insensitive.

**Comment by kingthorin** [comment]
Should the tag also cover other JSON'ish types?
https://github.com/zaproxy/zaproxy/blob/02b5c4f031e8cd6a741fa023310b68dd47d4dc7d/docker/scripts/scripts/httpsender/Alert_on_Unexpected_Content_Types.js#L12-L25

**Comment by kingthorin** [comment]
@psiinon @thc202 @ricekot [☝](https://github.com/zaproxy/zaproxy/issues/8399#issuecomment-2013304425) thoughts?

**Comment by kingthorin** [comment]
This was discussed via IRC. The decision was to not modify the existing pattern at this point. Others can be added, initially disabled.

Without further evidence of something actually misbehaving this issue will be closed.

**Comment by roisec** [comment]
Hi 
I have several examples of HTTP responses with content-type: application/json, and there is no JSON tag on the request in ZAP.

**Comment by kingthorin** [comment]
That's not evidence. Unless you can provide the details to back it up there's nothing we can do for you.

**Comment by roisec** [comment]
![image](https://github.com/zaproxy/zaproxy/assets/68992144/b8ed2d2e-ca1d-4599-bcc3-b6db0337ccf7)
In the following screenshots you are able to see some requests tagged by JSON.
When I send the same request again manually for this use case it works fine.
![image](https://github.com/zaproxy/zaproxy/assets/68992144/f0ea4cf8-2dfb-4535-a6ca-bcead3f45920)

**Comment by kingthorin** [comment]
As discussed above the existing pattern won't be updated in case anyone is depending on it in the current form. However there is already a PR in flight adding a json_extended pattern which would cover more circumstances.

**Comment by roisec** [comment]
@kingthorin thanks for your response,
Do you mean that there is no current solution for this bug?
Because this form does not cover tag JSON requests properly. (I can't depend on this)

**Comment by roisec** [comment]
Can I edit the regex currently?if yes,where?

**Comment by kingthorin** [comment]
Yes: https://www.zaproxy.org/docs/desktop/ui/dialogs/options/pscan/
Yes.

**Comment by roisec** [comment]
Ok, I checked this and configured my regex, and still, there is no automatic assignment of the relevant tag.
It's works only if I manually send the same request but not detected automatically by proxy requests.
@kingthorin  wdyt?

**Comment by roisec** [comment]
I created the tag regex and then generated the traffic .
if you want I able to share with you an screen recording

**Comment by kingthorin** [comment]
🤷‍♂️ There's a ton of variables here. Your regex could be wrong/poor, you might not be waiting for the passive scan queue, you might have "only in scope" on, etc. Unless you can provide the tag/regex and a public example of it not working then we can't recreate/troubleshoot.

**Comment by roisec** [comment]
The regex work perfect ,when I do the same request manually through zap the tag works properly 
Tomorrow I will send a screen recording that proves it

**Comment by kingthorin** [comment]
Make sure it includes the rule config and passive scan settings. Plus the message details of what is tagged and what's not.

**Comment by thc202** [comment]
https://github.com/zaproxy/zaproxy/issues/8399#issuecomment-2047441906

The existing pattern is not a/the problem anyway, it already handles the case `application/json; charset=utf-8`.

**Comment by roisec** [comment]
Let's see the video,
My tag and JSON tag are recognized only in manual, but through the proxy it's not detected automatically.
https://github.com/zaproxy/zaproxy/assets/68992144/f15b50b6-d225-4776-85d6-386de7603e3d

**Comment by thc202** [comment]
That's a problem with the UI not the passive scanner/regex.

**Comment by kingthorin** [comment]
Things you can try:
- On one/some of the entry(ies) after the passive scan queue has emptied (the Eyeball counter in the lower right), right click on an un-tagged entry and select "Manage History Tags" (do you see the tag there even if it wasn't shown in the table?).
- Run this standalone [community-script](https://github.com/zaproxy/community-scripts/blob/main/standalone/historySourceTagger.js) after the passive scan queue is empty.

**Comment by thc202** [comment]
There's no need to try, It's a bug.

**Comment by roisec** [comment]
@kingthorin I did 2 things you said, and still this is not tagged with JSON, also in "Manage History Tags".
Anyway, when the filter you implement in the PR  through the API ZAP (https://github.com/zaproxy/zaproxy/pull/8422), is should be working properly?

**Comment by kingthorin** [comment]
I haven't talked to @thc202 about the issue but he seems to have confirmed a UI issue, so no there will still be a piece outstanding it seems.
